File: web/src/lib/crypto/file.ts
import nacl from 'tweetnacl';
import type { EncryptedFileChunk } from '@/types/chat';

/**
 * 파일 청크 암호화 (nacl.box - XSalsa20-Poly1305)
 * 바이너리 직접 처리 (base64 변환 없음 — 33% 오버헤드 제거)
 */
export function encryptFileChunk(
  chunk: Uint8Array,
  sharedSecret: Uint8Array
): EncryptedFileChunk {
  const nonce = nacl.randomBytes(nacl.box.nonceLength);
  const ciphertext = nacl.box.after(chunk, nonce, sharedSecret);

  if (!ciphertext) {
    throw new Error('File chunk encryption failed');
  }

  return { ciphertext, nonce };
}

/**
 * 파일 청크 복호화 (nacl.box.open - XSalsa20-Poly1305)
 * 실패 시 null 반환 (변조 감지)
 */
export function decryptFileChunk(
  ciphertext: Uint8Array,
  nonce: Uint8Array,
  sharedSecret: Uint8Array
): Uint8Array | null {
  return nacl.box.open.after(ciphertext, nonce, sharedSecret);
}

File: mobile/lib/core/crypto/file.dart
import 'package:pinenacl/x25519.dart';

/// 암호화된 파일 청크
class EncryptedFileChunk {
  final Uint8List ciphertext;
  final Uint8List nonce;

  const EncryptedFileChunk({required this.ciphertext, required this.nonce});
}

/// 파일 청크 암호화 (web: nacl.box.after — 바이너리 직접 처리)
/// SecretBox는 내부적으로 crypto_box_afternm 사용 → 웹 호환
EncryptedFileChunk encryptFileChunk(Uint8List chunk, Uint8List sharedSecret) {
  final box = SecretBox(sharedSecret);
  final encrypted = box.encrypt(chunk);

  return EncryptedFileChunk(
    ciphertext: Uint8List.fromList(encrypted.cipherText),
    nonce: Uint8List.fromList(encrypted.nonce),
  );
}

/// 파일 청크 복호화
Uint8List? decryptFileChunk(
  Uint8List ciphertext,
  Uint8List nonce,
  Uint8List sharedSecret,
) {
  try {
    final box = SecretBox(sharedSecret);
    final encrypted = EncryptedMessage(
      nonce: nonce,
      cipherText: ciphertext,
    );
    return Uint8List.fromList(box.decrypt(encrypted));
  } catch (_) {
    return null;
  }
}
